logstash-core/lib/logstash/pipeline.rb

```diff
@@ -447,12 +447,14 @@ def worker_loop(batch_size, batch_delay)
       shutdown_requested |= signal.shutdown? # latch on shutdown signal
 
       batch = @filter_queue_client.read_batch # metrics are started in read_batch
```

Comment:
There is a timeout when waiting to fulfill a batch. Yields a short or empty batch. Also, due to conditionals or cancellations batches can become effectively empty. HTH